presto-main/src/main/java/io/prestosql/execution/SqlTask.java

```diff
@@ -340,32 +358,24 @@ private TaskInfo createTaskInfo(TaskHolder taskHolder)
                 needsPlan.get());
     }
 
-    public ListenableFuture<TaskStatus> getTaskStatus(TaskState callersCurrentState)
+    public ListenableFuture<TaskStatus> getTaskStatus(long callersCurrentVersion)
     {
-        requireNonNull(callersCurrentState, "callersCurrentState is null");
-
-        if (callersCurrentState.isDone()) {
+        if (callersCurrentVersion < taskStatusVersion.get() || taskHolderReference.get().isFinished()) {
```

Comment:
I'm pretty sure there is a race here.  The task could finish between this check and creating the listener below, and in that case the listener may never be notified.  If you call `createNewListener()` before this check, I believe it will eliminate the race.  Something like:
```java
CompletableFuture<?> listener = taskStatusVersionChange.createNewListener();
if (callersCurrentVersion < taskStatusVersion.get() || taskHolderReference.get().isFinished()) {
    listener.cancel(true);
    return immediateFuture(getTaskStatus());
}
return Futures.transform(listener, input -> getTaskStatus(), directExecutor());

Reply:
I've put both methods and notification method in `synchronized` section + a comment